File: sstvae/audio.py
"""Soundcard I/O: device enumeration and stream opening.

Everything here is optional at import time -- `sounddevice` (PortAudio)
is only needed by the live tools, so it is imported inside the functions
rather than at module scope. That keeps `import sstvae.audio` working on
a machine with no PortAudio, which matters because the settings dialog
wants to *report* that rather than fail to start.

The awkward part these helpers exist to hide: SSTVAE runs at 8 kHz, and
plenty of devices refuse that rate outright. PulseAudio and most modern
backends resample transparently, but ALSA hardware devices and some
Windows drivers do not, so both directions fall back to the device's
native rate with polyphase resampling on the way past.
"""

import sys
import threading
from dataclasses import dataclass
from math import gcd

import numpy as np

from .config import FS


class AudioUnavailable(RuntimeError):
    """PortAudio/sounddevice could not be loaded or found no devices."""


@dataclass(frozen=True)
class DeviceInfo:
    index: int
    name: str
    channels: int
    default_samplerate: float
    is_default: bool

    def label(self) -> str:
        star = " (default)" if self.is_default else ""
        return f"{self.index}: {self.name}{star}"


def _sd():
    try:
        import sounddevice as sd
    except Exception as e:  # ImportError, or PortAudio missing at load
        raise AudioUnavailable(
            f"sounddevice/PortAudio is not available: {e}\n"
            "Install the GUI extra (pip install -e .[gui]) and your platform's "
            "PortAudio library."
        ) from e
    return sd


def list_devices(kind: str) -> list[DeviceInfo]:
    """Input or output devices, in PortAudio index order.

    `kind` is "input" or "output". Devices with no channels in the
    requested direction are left out -- a card's playback side has no
    business appearing in the capture picker.
    """
    if kind not in ("input", "output"):
        raise ValueError(f"kind must be 'input' or 'output', got {kind!r}")
    sd = _sd()
    key = f"max_{kind}_channels"
    try:
        default_idx = sd.default.device[0 if kind == "input" else 1]
    except Exception:
        default_idx = None
    out = []
    for i, d in enumerate(sd.query_devices()):
        if d[key] < 1:
            continue
        out.append(
            DeviceInfo(
                index=i,
                name=d["name"],
                channels=d[key],
                default_samplerate=float(d["default_samplerate"]),
                is_default=(i == default_idx),
            )
        )
    if not out:
        raise AudioUnavailable(f"no {kind} devices found")
    return out


def _rate_fallback(device, kind: str) -> int:
    """The device's own preferred sample rate."""
    sd = _sd()
    info = sd.query_devices(device, kind)
    return int(round(info["default_samplerate"]))


def resample_ratio(src_rate: int, dst_rate: int) -> tuple[int, int]:
    """(up, down) for `scipy.signal.resample_poly` to convert audio at
    `src_rate` into audio at `dst_rate`.

    Spelled out with both rates named because the two directions are
    inverses of each other and sharing one "ratio to the device" helper
    between capture and playback silently got playback backwards: a
    transmission was decimated 48000->8000 instead of interpolated
    8000->48000, so 32 seconds of audio became 0.9 seconds of noise.
    """
    g = gcd(src_rate, dst_rate)
    return dst_rate // g, src_rate // g


class StreamResampler:
    """`resample_poly` for a stream that arrives in chunks.

    **Calling `resample_poly` on each callback chunk independently is
    wrong**, and wrong in a way that decodes rather than fails. It is an
    FIR polyphase filter, so an isolated chunk is zero-padded at both
    ends and every chunk boundary gets a filter transient; at 44.1 kHz
    into 8 kHz the filter is 8821 taps against ~186 output samples per
    chunk. On a real 66 s recording that cost **4.7 dB of SNR** (+2.4 dB
    one-shot, -2.3 dB per-chunk) and mangled the picture, while still
    syncing and reporting every frame received. `play()` sidesteps this
    by resampling the whole waveform up front; capture cannot.

    Per-chunk rounding is the second half of the bug: each chunk's output
    length is rounded up independently, so the stream gains samples --
    684 of them over that same recording, a 0.13% clock error the timing
    tracker then has to fight.

    This keeps `pad` input samples of context on *each* side of every
    block it emits, and only emits whole multiples of `down` input
    samples, so the output is sample-for-sample what one-shot resampling
    of the whole stream would have produced. Costs `2 * pad` input
    samples of latency, ~20 ms at 44.1 kHz.
    """

    def __init__(self, up: int, down: int):
        self.up = up
        self.down = down
        # scipy's default filter is 20*max(up, down)+1 taps in the
        # upsampled domain; this is its span measured in *input* samples.
        span = -(-(20 * max(up, down) + 1) // up)
        # Round up to a whole number of `down`, so the samples skipped at
        # the head are an exact integer and no phase error accumulates.
        self.pad = -(-span // down) * down
        self._buf = np.zeros(self.pad, dtype=np.float64)

    def __call__(self, chunk: np.ndarray) -> np.ndarray:
        from scipy.signal import resample_poly

        self._buf = np.concatenate(
            [self._buf, np.asarray(chunk, dtype=np.float64).reshape(-1)])
        usable = len(self._buf) - 2 * self.pad
        n = (usable // self.down) * self.down if usable > 0 else 0
        if n <= 0:
            return np.empty(0)
        block = self._buf[: 2 * self.pad + n]
        y = resample_poly(block, self.up, self.down)
        skip = self.pad * self.up // self.down
        take = n * self.up // self.down
        self._buf = self._buf[n:]
        return y[skip:skip + take]


def open_input_stream(device, ring, samplerate: int = FS, on_error=None):
    """Open an InputStream feeding `ring` (a `sstvae.rx.RingBuffer`).

    Returns (stream, actual_rate). Tries the requested rate directly
    first; falls back to the device's default rate with per-chunk
    polyphase resampling if that is rejected.
    """
    sd = _sd()
    report = on_error or (lambda msg: print(msg, file=sys.stderr))

    def make_callback(resample_fn=None):
        def callback(indata, frames, time_info, status):
            if status:
                report(f"[audio in] {status}")
            mono = indata[:, 0] if indata.ndim > 1 else indata
            ring.write(resample_fn(mono) if resample_fn else mono)

        return callback

    try:
        stream = sd.InputStream(
            samplerate=samplerate, channels=1, dtype="float32",
            device=device, callback=make_callback(),
        )
        stream.start()
        return stream, samplerate
    except Exception as e:
        native = _rate_fallback(device, "input")
        # Capture direction: the device hands us `native`, we want `samplerate`.
        up, down = resample_ratio(native, samplerate)
        report(
            f"[audio in] {samplerate} Hz rejected ({e}); falling back to "
            f"device default {native} Hz with resampling"
        )
        # Stateful across callbacks -- see StreamResampler. A bare
        # `resample_poly` per chunk decodes to a mangled picture.
        stream = sd.InputStream(
            samplerate=native, channels=1, dtype="float32", device=device,
            callback=make_callback(StreamResampler(up, down)),
        )
        stream.start()
        return stream, native


def play(device, samples: np.ndarray, samplerate: int = FS, on_progress=None,
         should_stop=None, on_error=None) -> bool:
    """Play `samples` (float, at `samplerate`) to `device`, blocking
    until it has finished or been stopped.

    `on_progress(fraction)` is called from the audio callback as the
    buffer drains, `should_stop()` is polled to allow cancellation.
    Returns True if playback completed, False if it was stopped early.

    Resamples up front rather than per-chunk when the device rejects the
    requested rate -- unlike capture, the whole waveform is already in
    hand, and one clean resample avoids polyphase edge effects at every
    block boundary.
    """
    sd = _sd()
    report = on_error or (lambda msg: print(msg, file=sys.stderr))
    x = np.asarray(samples, dtype=np.float32).reshape(-1)

    try:
        stream_rate = samplerate
        sd.check_output_settings(device=device, samplerate=samplerate,
                                 channels=1, dtype="float32")
    except Exception as e:
        from scipy.signal import resample_poly

        native = _rate_fallback(device, "output")
        # Playback direction: we hold `samplerate`, the device wants
        # `native` -- the opposite of the capture case above.
        up, down = resample_ratio(samplerate, native)
        report(
            f"[audio out] {samplerate} Hz rejected ({e}); resampling to "
            f"device default {native} Hz"
        )
        x = resample_poly(x, up, down).astype(np.float32)
        stream_rate = native

    total = len(x)
    pos = 0
    done = threading.Event()

    def callback(outdata, frames, time_info, status):
        nonlocal pos
        if status:
            report(f"[audio out] {status}")
        if should_stop is not None and should_stop():
            outdata[:] = 0
            raise sd.CallbackAbort
        n = min(frames, total - pos)
        outdata[:n, 0] = x[pos : pos + n]
        if n < frames:
            outdata[n:, 0] = 0
        pos += n
        if on_progress is not None:
            on_progress(pos / total if total else 1.0)
        if pos >= total:
            raise sd.CallbackStop

    stream = sd.OutputStream(
        samplerate=stream_rate, channels=1, dtype="float32",
        device=device, callback=callback, finished_callback=done.set,
    )
    with stream:
        # Poll rather than block forever: a device that stops calling
        # back (USB unplugged mid-transmission) must not wedge the
        # caller, because the caller is what drops PTT.
        while not done.wait(0.1):
            if should_stop is not None and should_stop():
                break
            if not stream.active:
                break
    return pos >= total

File: native/core/audio/audio.cpp
#include "audio/audio.hpp"

#include <algorithm>
#include <cmath>
#include <cstring>
#include <limits>
#include <numeric>
#include <stdexcept>

#include "dsp/dsp.hpp"

namespace sstvae::audio {

namespace {

// Ceiling division for positive values.
std::size_t ceil_div(std::size_t a, std::size_t b) { return (a + b - 1) / b; }

struct Scaling {
    double scale;   // divisor when reading
    double offset;  // zero point
    double lo;      // clip bounds in the integer domain
    double hi;
};

Scaling scaling(SampleFormat f) {
    switch (f) {
        case SampleFormat::Float:
            return {1.0, 0.0, -1.0, 1.0};
        case SampleFormat::Int16:
            return {32768.0, 0.0, -32768.0, 32767.0};
        case SampleFormat::Int32:
            return {2147483648.0, 0.0, -2147483648.0, 2147483647.0};
        case SampleFormat::UInt8:
            return {128.0, 128.0, 0.0, 255.0};
    }
    throw std::invalid_argument("unknown sample format");
}

// Read one sample of `fmt` from `p`, as a raw (unscaled) double.
double read_raw(const std::byte* p, SampleFormat f) {
    switch (f) {
        case SampleFormat::Float: {
            float v;
            std::memcpy(&v, p, sizeof v);
            return static_cast<double>(v);
        }
        case SampleFormat::Int16: {
            std::int16_t v;
            std::memcpy(&v, p, sizeof v);
            return static_cast<double>(v);
        }
        case SampleFormat::Int32: {
            std::int32_t v;
            std::memcpy(&v, p, sizeof v);
            return static_cast<double>(v);
        }
        case SampleFormat::UInt8:
            return static_cast<double>(std::to_integer<std::uint8_t>(*p));
    }
    throw std::invalid_argument("unknown sample format");
}

void write_raw(std::byte* p, SampleFormat f, double v) {
    switch (f) {
        case SampleFormat::Float: {
            const auto x = static_cast<float>(v);
            std::memcpy(p, &x, sizeof x);
            return;
        }
        case SampleFormat::Int16: {
            const auto x = static_cast<std::int16_t>(v);
            std::memcpy(p, &x, sizeof x);
            return;
        }
        case SampleFormat::Int32: {
            const auto x = static_cast<std::int32_t>(v);
            std::memcpy(p, &x, sizeof x);
            return;
        }
        case SampleFormat::UInt8:
            *p = static_cast<std::byte>(static_cast<std::uint8_t>(v));
            return;
    }
    throw std::invalid_argument("unknown sample format");
}

std::string lowered(std::string_view s) {
    std::string out(s);
    std::transform(out.begin(), out.end(), out.begin(),
                   [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
    return out;
}

}  // namespace

Ratio resample_ratio(int src_rate, int dst_rate) {
    if (src_rate <= 0 || dst_rate <= 0) {
        throw std::invalid_argument("resample_ratio: rates must be positive");
    }
    const int g = std::gcd(src_rate, dst_rate);
    return {dst_rate / g, src_rate / g};
}

StreamResampler::StreamResampler(int up, int down) : up_(up), down_(down) {
    if (up <= 0 || down <= 0) {
        throw std::invalid_argument("StreamResampler: up and down must be positive");
    }
    // The default filter is 20*max(up, down)+1 taps in the *upsampled*
    // domain; this is its span measured in input samples.
    const std::size_t span =
        ceil_div(static_cast<std::size_t>(20 * std::max(up, down) + 1),
                 static_cast<std::size_t>(up));
    // Rounded up to a whole number of `down`, so the samples skipped at
    // the head are an exact integer and no phase error accumulates.
    pad_ = ceil_div(span, static_cast<std::size_t>(down)) * static_cast<std::size_t>(down);
    buf_.assign(pad_, 0.0);
}

std::vector<double> StreamResampler::operator()(std::span<const double> chunk) {
    buf_.insert(buf_.end(), chunk.begin(), chunk.end());

    const std::size_t d = static_cast<std::size_t>(down_);
    const std::size_t usable = buf_.size() > 2 * pad_ ? buf_.size() - 2 * pad_ : 0;
    const std::size_t n = (usable / d) * d;
    if (n == 0) return {};

    const std::vector<double> y = dsp::resample_poly(
        std::span<const double>(buf_.data(), 2 * pad_ + n), up_, down_, &filter_taps_);
    const std::size_t skip = pad_ * static_cast<std::size_t>(up_) / d;
    const std::size_t take = n * static_cast<std::size_t>(up_) / d;

    buf_.erase(buf_.begin(), buf_.begin() + static_cast<std::ptrdiff_t>(n));
    return std::vector<double>(y.begin() + static_cast<std::ptrdiff_t>(skip),
                               y.begin() + static_cast<std::ptrdiff_t>(skip + take));
}

std::optional<SampleFormat> sample_format_from_name(std::string_view name) {
    if (name == "Float") return SampleFormat::Float;
    if (name == "Int16") return SampleFormat::Int16;
    if (name == "Int32") return SampleFormat::Int32;
    if (name == "UInt8") return SampleFormat::UInt8;
    return std::nullopt;
}

const char* sample_format_name(SampleFormat f) {
    switch (f) {
        case SampleFormat::Float: return "Float";
        case SampleFormat::Int16: return "Int16";
        case SampleFormat::Int32: return "Int32";
        case SampleFormat::UInt8: return "UInt8";
    }
    return "?";
}

int bytes_per_sample(SampleFormat f) {
    switch (f) {
        case SampleFormat::Float: return 4;
        case SampleFormat::Int16: return 2;
        case SampleFormat::Int32: return 4;
        case SampleFormat::UInt8: return 1;
    }
    return 0;
}

std::vector<double> bytes_to_mono(std::span<const std::byte> raw, SampleFormat fmt,
                                  int channels) {
    if (channels < 1) throw std::invalid_argument("bytes_to_mono: channels < 1");
    const auto bps = static_cast<std::size_t>(bytes_per_sample(fmt));
    const std::size_t frame = bps * static_cast<std::size_t>(channels);
    // A trailing partial frame is dropped: keeping it would misalign
    // every sample after it, which sounds like noise rather than an
    // error.
    const std::size_t frames = raw.size() / frame;
    const Scaling s = scaling(fmt);

    std::vector<double> out(frames);
    for (std::size_t i = 0; i < frames; ++i) {
        double acc = 0.0;
        for (int c = 0; c < channels; ++c) {
            acc += read_raw(raw.data() + i * frame + static_cast<std::size_t>(c) * bps,
                            fmt);
        }
        out[i] = (acc / channels - s.offset) / s.scale;
    }
    return out;
}

std::vector<std::byte> mono_to_bytes(std::span<const double> x, SampleFormat fmt,
                                     int channels) {
    if (channels < 1) throw std::invalid_argument("mono_to_bytes: channels < 1");
    const auto bps = static_cast<std::size_t>(bytes_per_sample(fmt));
    const std::size_t frame = bps * static_cast<std::size_t>(channels);
    const Scaling s = scaling(fmt);

    std::vector<std::byte> out(x.size() * frame);
    for (std::size_t i = 0; i < x.size(); ++i) {
        double v;
        if (fmt == SampleFormat::Float) {
            v = std::clamp(x[i], -1.0, 1.0);
        } else {
            // Scale by `scale - 1` and clip in the integer domain, so
            // exactly +1.0 lands on the positive full-scale code rather
            // than wrapping to the negative one. Half-to-even rounding,
            // like numpy's.
            v = std::clamp(std::nearbyint(x[i] * (s.scale - 1.0) + s.offset), s.lo, s.hi);
        }
        for (int c = 0; c < channels; ++c) {
            write_raw(out.data() + i * frame + static_cast<std::size_t>(c) * bps, fmt, v);
        }
    }
    return out;
}

std::optional<std::size_t> match_device(std::span<const std::string> descriptions,
                                        std::string_view wanted) {
    if (wanted.empty()) return std::nullopt;
    for (std::size_t i = 0; i < descriptions.size(); ++i) {
        if (descriptions[i] == wanted) return i;
    }
    const std::string low = lowered(wanted);
    std::optional<std::size_t> hit;
    for (std::size_t i = 0; i < descriptions.size(); ++i) {
        if (lowered(descriptions[i]).find(low) == std::string::npos) continue;
        if (hit) return std::nullopt;  // ambiguous: prefer the default
        hit = i;
    }
    return hit;
}

}  // namespace sstvae::audio
